File: BackEnd/src/domain/entities/PlantSpecies.ts
/**
 * @file PlantSpecies.ts
 * @description Entidad de dominio que representa una especie de planta (catálogo compartido).
 * Puede ser creada por usuarios (propuesta) y aprobada por admins para ser pública.
 * @module Plants
 * @layer Domain
 */

import type { LightNeed } from './Plant.js';

/**
 * Rango de temperatura óptima para la especie.
 */
export interface TemperatureRange {
  /** Temperatura mínima en grados Celsius */
  min: number;
  /** Temperatura máxima en grados Celsius */
  max: number;
}

/**
 * Ajuste estacional de la frecuencia de riego de la especie.
 * Los valores son multiplicadores sobre wateringFrequency:
 *   < 1.0 → regar más frecuentemente (p.ej. 0.7 = 30% más frecuente)
 *   > 1.0 → regar menos frecuentemente (p.ej. 1.5 = 50% menos frecuente)
 * Primavera y otoño no se ajustan (factor = 1.0).
 */
export interface SeasonalWateringAdjustment {
  /** Multiplicador de riego en verano (jun–ago en hemisferio norte). */
  summer?: number;
  /** Multiplicador de riego en invierno (dic–feb en hemisferio norte). */
  winter?: number;
}

/**
 * Requisitos de cuidado base de la especie.
 * Sirven como valores por defecto cuando el usuario crea una planta de esta especie.
 */
export interface CareRequirements {
  /** Frecuencia de riego recomendada en días */
  wateringDays: number;
  /** Necesidad de luz */
  lightNeed: LightNeed;
  /** Rango de temperatura óptima (opcional) */
  temperatureRange?: TemperatureRange;
}

/**
 * Entrada del historial de auditoría de cambios en la especie.
 */
export interface AuditEntry {
  /** Fecha del cambio */
  date: Date;
  /** Id del usuario que realizó el cambio */
  userId: string;
  /** Descripción resumida de los cambios */
  changes: string;
}

/**
 * Entidad de dominio PlantSpecies.
 * Representa una especie del catálogo compartido de la aplicación.
 */
export class PlantSpecies {
  /** Identificador único */
  readonly id: string;

  /** Nombre común de la especie */
  readonly name: string;

  /** Nombre científico (binomial) */
  readonly scientificName: string;

  /** URL de la imagen representativa en Cloudinary (nullable — puede no tener imagen asignada) */
  readonly image?: string;

  /** Requisitos de cuidado base */
  readonly careRequirements: CareRequirements;

  /** Climas con los que es compatible (ej: "Mediterráneo", "Tropical", "Árido") */
  readonly climateCompatibility: string[];

  /** Consejos de cuidado */
  readonly tips: string[];

  /** Id del usuario que propuso la especie (undefined si es del sistema) */
  readonly createdBy?: string;

  /** Si true, la especie es visible para todos los usuarios */
  readonly isPublic: boolean;

  /** Indica si la especie requiere poda anual */
  readonly requiresPruning?: boolean;

  /** Meses recomendados de poda (1 = enero, 12 = diciembre). Solo aplica si requiresPruning es true */
  readonly pruningMonths?: number[];

  /** Indica si la especie produce frutos o cosecha */
  readonly produceFruit?: boolean;

  /** Meses del año en los que se puede cosechar (1 = enero, 12 = diciembre). Solo aplica si produceFruit es true */
  readonly harvestMonths?: number[];

  /** Ajuste estacional de la frecuencia de riego (opcional). */
  readonly seasonalWateringAdjustment?: SeasonalWateringAdjustment;

  /** Precipitación mínima en mm para considerar la planta regada por lluvia (opcional). Default efectivo: 5mm. */
  readonly minRainfallMm?: number;

  /** Cantidad de agua en litros que necesita la planta en cada riego (opcional, informativo). */
  readonly waterLitersPerWatering?: number;

  /** Historial de cambios realizados por admins */
  readonly auditHistory: AuditEntry[];

  /** Fecha de creación */
  readonly createdAt: Date;

  /** Fecha de última modificación */
  readonly updatedAt: Date;

  /** Fecha de borrado lógico */
  readonly deletedAt?: Date | null;

  constructor(params: {
    id: string;
    name: string;
    scientificName: string;
    image?: string;
    careRequirements: CareRequirements;
    climateCompatibility: string[];
    tips: string[];
    createdBy?: string;
    isPublic: boolean;
    requiresPruning?: boolean;
    pruningMonths?: number[];
    produceFruit?: boolean;
    harvestMonths?: number[];
    seasonalWateringAdjustment?: SeasonalWateringAdjustment;
    minRainfallMm?: number;
    waterLitersPerWatering?: number;
    auditHistory: AuditEntry[];
    createdAt: Date;
    updatedAt: Date;
    deletedAt?: Date | null;
  }) {
    this.id = params.id;
    this.name = params.name;
    this.scientificName = params.scientificName;
    this.image = params.image;
    this.careRequirements = params.careRequirements;
    this.climateCompatibility = params.climateCompatibility;
    this.tips = params.tips;
    this.createdBy = params.createdBy;
    this.isPublic = params.isPublic;
    this.requiresPruning = params.requiresPruning;
    this.pruningMonths = params.pruningMonths;
    this.produceFruit = params.produceFruit;
    this.harvestMonths = params.harvestMonths;
    this.seasonalWateringAdjustment = params.seasonalWateringAdjustment;
    this.minRainfallMm = params.minRainfallMm;
    this.waterLitersPerWatering = params.waterLitersPerWatering;
    this.auditHistory = params.auditHistory;
    this.createdAt = params.createdAt;
    this.updatedAt = params.updatedAt;
    this.deletedAt = params.deletedAt;
  }

  /**
   * Compara esta especie con otra y devuelve los campos que difieren.
   * Útil para el panel de revisión de propuestas de nuevas especies.
   *
   * @param other — Especie a comparar.
   * @returns Objeto con los campos que tienen valores distintos.
   */
  diffWith(other: PlantSpecies): Partial<Record<keyof PlantSpecies, { current: unknown; proposed: unknown }>> {
    const diff: Partial<Record<keyof PlantSpecies, { current: unknown; proposed: unknown }>> = {};

    const fields: Array<keyof PlantSpecies> = [
      'name', 'scientificName', 'image', 'careRequirements',
      'climateCompatibility', 'tips', 'isPublic',
    ];

    for (const field of fields) {
      const a = JSON.stringify(this[field]);
      const b = JSON.stringify(other[field]);
      if (a !== b) {
        diff[field] = { current: this[field], proposed: other[field] };
      }
    }

    return diff;
  }

  /**
   * Indica si la especie es una propuesta pendiente de aprobación.
   */
  get isPendingApproval(): boolean {
    return !this.isPublic && !!this.createdBy;
  }
}

File: FrontEnd/lib/domain/entities/plant_species.dart
/// @file plant_species.dart
/// @description Entidad de dominio PlantSpecies. Representa una especie del catálogo.
/// Objeto puro de Dart, sin dependencias de Flutter ni paquetes externos.
/// Los mappers convierten PlantSpeciesModel ↔ PlantSpecies.
/// @module Plants
/// @layer Domain
library;

// ═══════════════════════════════════════════════════════════════════════════════
// PLANT SPECIES ENTITY
// ═══════════════════════════════════════════════════════════════════════════════

/// Entidad de dominio que representa una especie de planta del catálogo.
///
/// Todos los campos son inmutables (final). Usar [copyWith] para actualizaciones.
class PlantSpecies {
  const PlantSpecies({
    required this.id,
    required this.name,
    required this.scientificName,
    required this.image,
    required this.careRequirements,
    required this.climateCompatibility,
    required this.tips,
    required this.isPublic,
    required this.createdBy,
    required this.createdAt,
    required this.updatedAt,
    this.requiresPruning,
    this.pruningMonths,
    this.produceFruit,
    this.harvestMonths,
    this.seasonalWateringAdjustment,
    this.minRainfallMm,
    this.waterLitersPerWatering,
  });

  /// Identificador único de la especie (MongoDB ObjectId como String).
  final String id;

  /// Nombre común de la especie (p.ej. "Monstera").
  final String name;

  /// Nombre científico (p.ej. "Monstera deliciosa").
  final String scientificName;

  /// URL de la imagen representativa de la especie, o null si no tiene imagen asignada.
  final String? image;

  /// Requisitos de cuidado estándar de la especie.
  final SpeciesCareRequirements careRequirements;

  /// Climas compatibles con la especie (p.ej. ['tropical', 'subtropical']).
  final List<String> climateCompatibility;

  /// Consejos de cuidado y curiosidades.
  final List<String> tips;

  /// true si la especie está aprobada y visible en el catálogo público.
  final bool isPublic;

  /// ID del usuario que propuso la especie.
  final String createdBy;

  /// Fecha de creación del registro.
  final DateTime createdAt;

  /// Fecha de última actualización.
  final DateTime updatedAt;

  /// Indica si la especie requiere poda anual.
  final bool? requiresPruning;

  /// Meses recomendados de poda (1 = enero, 12 = diciembre). Solo aplica si [requiresPruning] es true.
  final List<int>? pruningMonths;

  /// Indica si la especie produce frutos o cosecha.
  final bool? produceFruit;

  /// Meses del año en los que se puede cosechar (1 = enero, 12 = diciembre). Solo aplica si [produceFruit] es true.
  final List<int>? harvestMonths;

  /// Ajuste estacional de la frecuencia de riego (opcional).
  final SeasonalWateringAdjustment? seasonalWateringAdjustment;

  /// Precipitación mínima en mm para considerar la planta regada por lluvia (opcional).
  final double? minRainfallMm;

  /// Cantidad de agua en litros por riego (opcional, informativo).
  final double? waterLitersPerWatering;

  // ─── Helpers ─────────────────────────────────────────────────────────────────

  /// true si la especie está pendiente de aprobación.
  bool get isPendingApproval => !isPublic;

  /// Frecuencia de riego estándar de la especie en días.
  int get standardWateringDays => careRequirements.wateringDays;

  // ─── copyWith ────────────────────────────────────────────────────────────────

  PlantSpecies copyWith({
    String?                  id,
    String?                  name,
    String?                  scientificName,
    String?                  image,  // null preserva el valor actual; no fuerza null explícito
    SpeciesCareRequirements? careRequirements,
    List<String>?            climateCompatibility,
    List<String>?            tips,
    bool?                    isPublic,
    String?                  createdBy,
    DateTime?                createdAt,
    DateTime?                updatedAt,
    bool?                         requiresPruning,
    List<int>?                    pruningMonths,
    bool?                         produceFruit,
    List<int>?                    harvestMonths,
    SeasonalWateringAdjustment?   seasonalWateringAdjustment,
    double?                       minRainfallMm,
    double?                       waterLitersPerWatering,
  }) {
    return PlantSpecies(
      id:                          id                          ?? this.id,
      name:                        name                        ?? this.name,
      scientificName:              scientificName              ?? this.scientificName,
      image:                       image                       ?? this.image,
      careRequirements:            careRequirements            ?? this.careRequirements,
      climateCompatibility:        climateCompatibility        ?? this.climateCompatibility,
      tips:                        tips                        ?? this.tips,
      isPublic:                    isPublic                    ?? this.isPublic,
      createdBy:                   createdBy                   ?? this.createdBy,
      createdAt:                   createdAt                   ?? this.createdAt,
      updatedAt:                   updatedAt                   ?? this.updatedAt,
      requiresPruning:             requiresPruning             ?? this.requiresPruning,
      pruningMonths:                pruningMonths                ?? this.pruningMonths,
      produceFruit:                produceFruit                ?? this.produceFruit,
      harvestMonths:               harvestMonths               ?? this.harvestMonths,
      seasonalWateringAdjustment:  seasonalWateringAdjustment  ?? this.seasonalWateringAdjustment,
      minRainfallMm:               minRainfallMm               ?? this.minRainfallMm,
      waterLitersPerWatering:      waterLitersPerWatering      ?? this.waterLitersPerWatering,
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) || (other is PlantSpecies && other.id == id);

  @override
  int get hashCode => id.hashCode;

  @override
  String toString() => 'PlantSpecies(id: $id, name: $name)';
}

// ═══════════════════════════════════════════════════════════════════════════════
// SPECIES CARE REQUIREMENTS
// ═══════════════════════════════════════════════════════════════════════════════

/// Requisitos de cuidado estándar definidos en el catálogo para una especie.
class SpeciesCareRequirements {
  const SpeciesCareRequirements({
    required this.wateringDays,
    required this.lightNeed,
    this.temperatureRange,
  });

  /// Frecuencia de riego estándar en días (p.ej. 7 = semanal).
  final int wateringDays;

  /// Necesidad de luz: 'Low', 'Medium' o 'High'.
  final String lightNeed;

  /// Rango de temperatura óptima (°C), o null si no está especificado.
  final SpeciesTemperatureRange? temperatureRange;
}

// ═══════════════════════════════════════════════════════════════════════════════
// SPECIES TEMPERATURE RANGE
// ═══════════════════════════════════════════════════════════════════════════════

/// Rango de temperatura óptima para la especie (en °C).
class SpeciesTemperatureRange {
  const SpeciesTemperatureRange({required this.min, required this.max});

  /// Temperatura mínima óptima en °C.
  final double min;

  /// Temperatura máxima óptima en °C.
  final double max;
}

// ═══════════════════════════════════════════════════════════════════════════════
// SEASONAL WATERING ADJUSTMENT
// ═══════════════════════════════════════════════════════════════════════════════

/// Ajuste estacional de la frecuencia de riego de la especie.
///
/// Los valores son multiplicadores sobre [careRequirements.wateringDays]:
///   < 1.0 → regar más frecuentemente (p.ej. 0.7 = 30% más frecuente en verano)
///   > 1.0 → regar menos frecuentemente (p.ej. 1.5 = 50% menos en invierno)
/// Primavera y otoño no se ajustan (factor = 1.0).
class SeasonalWateringAdjustment {
  const SeasonalWateringAdjustment({this.summer, this.winter});

  /// Multiplicador de riego en verano (jun–ago). Null = sin ajuste.
  final double? summer;

  /// Multiplicador de riego en invierno (dic–feb). Null = sin ajuste.
  final double? winter;
}
